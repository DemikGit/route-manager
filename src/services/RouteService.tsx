import * as React from 'react';
import { GoogleMap, InfoWindow, Marker, Polyline } from 'react-google-maps';

export interface IMarker {
  id: string,
  name: string,
  draggable: boolean,
  showInfoBox: boolean,
  onDrag: (event: google.maps.MouseEvent, id: string) => void,
  position: {
    lat: number,
    lng: number,
  }
}

export class RouteService {
  /* tslint:disable variable-name*/
  private _markers: IMarker[];
  private _mapRef: React.RefObject<GoogleMap>;
  private _componentUpdate: () => void;
  /* tslint:enable variable-name */

  private defaultPoint: IMarker;
  private polylineOptions: {
      strokeColor: string,
      strokeOpacity: number,
      strokeWeight: number,
  };

  public set updateComponent(componentUpdate: () => void) {
    this._componentUpdate = componentUpdate;
  }

  public get markers() : IMarker[] {
    return this._markers;
  }

  public set mapRef(mapRef: React.RefObject<GoogleMap>) {
    this._mapRef = mapRef;
  }

  constructor() {
    this._markers = [];
    this.polylineOptions = {
      strokeColor: '#4bdded',
      strokeOpacity: 0.85,
      strokeWeight: 4,
    };
    this.defaultPoint = {
      draggable: true,
      id: '',
      name: '',
      onDrag: (event: google.maps.MouseEvent, id: string) => {
        this.onDrag(event, id);
      },
      position: {
        lat: 0,
        lng: 0,
      },
      showInfoBox: false,
    };
  }

  public swapPoints = (srcId: string, targetId: string): void => {
    if(srcId !== targetId) {
      const newMarkers = this._markers;

      const srcMarkerIndex = this._markers.findIndex((marker: IMarker) => {
        return marker.id === srcId;
      });
      const srcMarker = this._markers[srcMarkerIndex];

      const targetMarkerIndex = this._markers.findIndex((marker: IMarker) => {
        return marker.id === targetId;
      });
      const targetMarker = this._markers[targetMarkerIndex];

      newMarkers[targetMarkerIndex] = srcMarker;
      newMarkers[srcMarkerIndex] = targetMarker;

      this._markers = newMarkers;
      this._componentUpdate();
    }
  }

  public onDrag = (event: google.maps.MouseEvent, id: string): void => {
    const { lat, lng } = event.latLng;

    const markerTarget: IMarker | undefined =
      this._markers.find((point: IMarker) => {
        return point.id.includes(id);
    });

    if (markerTarget) {
      markerTarget.position = { lat: lat() , lng: lng() };
    }

    this._componentUpdate();
  }

  public onPointDelete = (id: string): void => {
    this._markers = this._markers.filter((point: IMarker) => {
      return !point.id.includes(id);
    });

    this._componentUpdate();
  }

  public onAddPoint = (name: string): void => {
    if (this._mapRef.current) {

      const {
        lat: centerLat,
        lng: centerLng,
      } = this._mapRef.current.getCenter();

      this._markers.push({
        ...this.defaultPoint,
        id: this.uuidV4(),
        name,
        position: { lat: centerLat(), lng: centerLng() },
      });

      this._componentUpdate();
    }
  }

  public render = (): JSX.Element[] => {
    return this.renderMarkers().concat(this.renderPolyline());
  }

  private onToggleInfoBox = (id: string): void => {
    const targetMarker: IMarker | undefined = this._markers.find(
      (marker: IMarker): boolean => {
        return marker.id === id;
      }
    );

    if( targetMarker ) {
      targetMarker.showInfoBox = !targetMarker.showInfoBox;
    }

    this._componentUpdate();
  }

  private renderMarkers = (): JSX.Element[] => {
    const markers: JSX.Element[] = [];

    this._markers.forEach((marker) => {
      const { onDrag, draggable, position, id } = marker;
      const extendedDrag = ( event: google.maps.MouseEvent) => {
        onDrag(event, id);
      };

      const extendedToggleInfoBox = ( ) => {
        this.onToggleInfoBox(id);
      };

      markers.push(
        <Marker
          key={ id }
          onClick={ extendedToggleInfoBox }
          onDrag={ extendedDrag }
          draggable={ draggable }
          position={ position }
          labelAnchor={ new google.maps.Point(0, 0) }
        >
          { marker.showInfoBox && (
            <InfoWindow
              onCloseClick={ extendedToggleInfoBox }
            >
              <div>
                {`  ${ marker.name }`}
              </div>
            </InfoWindow>
          )}
        </Marker>
      );
    })

    return markers;
  }

  private renderPolyline = (): JSX.Element[] => {
    if (this._markers.length > 0) {
      const path: google.maps.LatLngLiteral[] = [];

      this._markers.forEach((marker: IMarker) => {
        path.push(marker.position);
      })

      return [
        <Polyline
          key={ this.uuidV4() }
          options={ this.polylineOptions }
          path={ path }
        />,
      ];
    }

    return [];
  }

  private uuidV4 = (): string => {
    // tslint:disable:no-bitwise
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
    // tslint:enable:no-bitwise
  }
}
