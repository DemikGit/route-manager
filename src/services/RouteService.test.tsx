import 'jest-enzyme';
import { RouteService } from './RouteService';

let accumulator = 0;

const mockUUID = jest.fn((): string => String(accumulator += 1));

it('add new marker', () => {
  const mockUpdateComponent = jest.fn();
  accumulator = 0;
  const routeService = new RouteService();

  routeService.updateComponent = mockUpdateComponent;
  routeService.uuidV4 = mockUUID;
  routeService.onAddPoint('Test Marker', { lat: 0, lng: 0 });
  expect(routeService.markers.length).toBeGreaterThan(0);
  expect(mockUpdateComponent).toHaveBeenCalledTimes(1);
  expect(routeService).toMatchSnapshot();
});

it('remove marker', () => {
  const mockUpdateComponent = jest.fn();
  const routeService = new RouteService();
  accumulator = 0;

  routeService.updateComponent = mockUpdateComponent;
  routeService.uuidV4 = mockUUID;

  routeService.onAddPoint('Test Marker', { lat: 0, lng: 0 });
  routeService.onAddPoint('Test Marker', { lat: 0, lng: 0 });
  routeService.onAddPoint('Test Marker', { lat: 0, lng: 0 });

  const lastMarkerId = routeService.markers[2].id;

  routeService.onPointDelete(lastMarkerId);

  const isMarkerExist = !!routeService.markers.find(
    marker => marker.id === lastMarkerId
  );

  expect(isMarkerExist).toEqual(false);
  expect(routeService.markers.length).toEqual(2);
  expect(mockUpdateComponent).toHaveBeenCalledTimes(4);
  expect(routeService).toMatchSnapshot();
});

it('change marker position when dragging', () => {
  const mockUpdateComponent = jest.fn();
  const newPointPosition = {
    lat: 15,
    lng: 15,
  };
  accumulator = 0;
  const routeService = new RouteService();
  routeService.updateComponent = mockUpdateComponent;
  routeService.uuidV4 = mockUUID;

  routeService.onAddPoint('Test Marker', { lat: 0, lng: 0 });
  const marker = routeService.markers[0];

  routeService.onDrag({ ...newPointPosition }, marker.id);

  expect(routeService.markers.length).toBeGreaterThan(0);
  expect(routeService.markers[0].position).toEqual(newPointPosition);
  expect(mockUpdateComponent).toHaveBeenCalledTimes(2);
  expect(routeService).toMatchSnapshot();
});

it('swap markers in list', () => {
  const mockUpdateComponent = jest.fn();
  accumulator = 0;
  const routeService = new RouteService();
  routeService.updateComponent = mockUpdateComponent;
  routeService.uuidV4 = mockUUID;

  routeService.onAddPoint('Test Marker 1', { lat: 0, lng: 0 });
  routeService.onAddPoint('Test Marker 2', { lat: 0, lng: 0 });
  routeService.onAddPoint('Test Marker 3', { lat: 0, lng: 0 });
  const lastMarkerId = routeService.markers[2].id;
  const firstMarkerId = routeService.markers[0].id;

  routeService.swapPoints(firstMarkerId, lastMarkerId);

  expect(routeService.markers[0].id).toEqual(lastMarkerId);
  expect(routeService.markers[2].id).toEqual(firstMarkerId);
  expect(routeService.markers.length).toEqual(3);
  expect(mockUpdateComponent).toHaveBeenCalledTimes(4);
  expect(routeService).toMatchSnapshot();
});

it('toggle markers babble', () => {
  const mockUpdateComponent = jest.fn();
  const routeService = new RouteService();
  routeService.updateComponent = mockUpdateComponent;
  accumulator = 0;
  routeService.uuidV4 = mockUUID;

  routeService.onAddPoint('Test Marker 1', { lat: 0, lng: 0 });
  const firstMarker = routeService.markers[0];

  expect(firstMarker.showInfoBox).toEqual(false);
  routeService.onToggleInfoBox(firstMarker.id);
  expect(firstMarker.showInfoBox).toEqual(true);

  expect(routeService.markers.length).toEqual(1);
  expect(mockUpdateComponent).toHaveBeenCalledTimes(2);
  expect(routeService).toMatchSnapshot();
});

it('test rendering JSX.Element', () => {
  const mockUpdateComponent = jest.fn();
  const routeService = new RouteService();
  class Point {
    /** The X coordinate */
    x: number;
    /** The Y coordinate */
    y: number;
    /** A point on a two-dimensional plane. */
    constructor(x: number, y: number) {
      this.x = x;
      this.y = y;
    }
    /** Compares two Points */
    equals(other: Point): boolean {
      return false;
    }
    /** Returns a string representation of this Point. */
    toString(): string {
      return '';
    }
  }

  const google = {
    maps: {
      Point
    },
  }

  global.window.google = google;

  routeService.updateComponent = mockUpdateComponent;
  routeService.uuidV4 = mockUUID;
  accumulator = 0;
  routeService.onAddPoint('Test Marker 1', { lat: 0, lng: 0 });
  routeService.onAddPoint('Test Marker 2', { lat: 0, lng: 0 });
  routeService.onAddPoint('Test Marker 3', { lat: 0, lng: 0 });

  const renderedElements = routeService.render();

  const markersCount = renderedElements.filter(element => {
    return element.type.name === 'Marker'
  }).length;
  const polylineCount = renderedElements.filter(element => {
    return element.type.name === 'Polyline'
  }).length;

  expect(routeService.markers.length).toEqual(3);
  expect(markersCount).toEqual(3);
  expect(polylineCount).toEqual(1);
  expect(mockUpdateComponent).toHaveBeenCalledTimes(3);
  expect(routeService).toMatchSnapshot();
});

