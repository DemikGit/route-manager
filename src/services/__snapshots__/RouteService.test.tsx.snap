// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add new marker 1`] = `
RouteService {
  "_componentUpdate": [MockFunction] {
    "calls": Array [
      Array [],
    ],
  },
  "_markers": Array [
    Object {
      "draggable": true,
      "id": "1",
      "name": "Test Marker",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
  ],
  "defaultPoint": Object {
    "draggable": true,
    "id": "",
    "name": "",
    "onDrag": [Function],
    "position": Object {
      "lat": 0,
      "lng": 0,
    },
    "showInfoBox": false,
  },
  "onAddPoint": [Function],
  "onDrag": [Function],
  "onPointDelete": [Function],
  "onToggleInfoBox": [Function],
  "polylineOptions": Object {
    "strokeColor": "#4bdded",
    "strokeOpacity": 0.85,
    "strokeWeight": 4,
  },
  "render": [Function],
  "renderMarkers": [Function],
  "renderPolyline": [Function],
  "swapPoints": [Function],
  "uuidV4": [MockFunction] {
    "calls": Array [
      Array [],
    ],
  },
}
`;

exports[`change marker position when dragging 1`] = `
RouteService {
  "_componentUpdate": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
    ],
  },
  "_markers": Array [
    Object {
      "draggable": true,
      "id": "1",
      "name": "Test Marker",
      "onDrag": [Function],
      "position": Object {
        "lat": 15,
        "lng": 15,
      },
      "showInfoBox": false,
    },
  ],
  "defaultPoint": Object {
    "draggable": true,
    "id": "",
    "name": "",
    "onDrag": [Function],
    "position": Object {
      "lat": 0,
      "lng": 0,
    },
    "showInfoBox": false,
  },
  "onAddPoint": [Function],
  "onDrag": [Function],
  "onPointDelete": [Function],
  "onToggleInfoBox": [Function],
  "polylineOptions": Object {
    "strokeColor": "#4bdded",
    "strokeOpacity": 0.85,
    "strokeWeight": 4,
  },
  "render": [Function],
  "renderMarkers": [Function],
  "renderPolyline": [Function],
  "swapPoints": [Function],
  "uuidV4": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
    ],
  },
}
`;

exports[`remove marker 1`] = `
RouteService {
  "_componentUpdate": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
      Array [],
    ],
  },
  "_markers": Array [
    Object {
      "draggable": true,
      "id": "1",
      "name": "Test Marker",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
    Object {
      "draggable": true,
      "id": "2",
      "name": "Test Marker",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
  ],
  "defaultPoint": Object {
    "draggable": true,
    "id": "",
    "name": "",
    "onDrag": [Function],
    "position": Object {
      "lat": 0,
      "lng": 0,
    },
    "showInfoBox": false,
  },
  "onAddPoint": [Function],
  "onDrag": [Function],
  "onPointDelete": [Function],
  "onToggleInfoBox": [Function],
  "polylineOptions": Object {
    "strokeColor": "#4bdded",
    "strokeOpacity": 0.85,
    "strokeWeight": 4,
  },
  "render": [Function],
  "renderMarkers": [Function],
  "renderPolyline": [Function],
  "swapPoints": [Function],
  "uuidV4": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
      Array [],
    ],
  },
}
`;

exports[`swap markers in list 1`] = `
RouteService {
  "_componentUpdate": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
      Array [],
    ],
  },
  "_markers": Array [
    Object {
      "draggable": true,
      "id": "3",
      "name": "Test Marker 3",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
    Object {
      "draggable": true,
      "id": "2",
      "name": "Test Marker 2",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
    Object {
      "draggable": true,
      "id": "1",
      "name": "Test Marker 1",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
  ],
  "defaultPoint": Object {
    "draggable": true,
    "id": "",
    "name": "",
    "onDrag": [Function],
    "position": Object {
      "lat": 0,
      "lng": 0,
    },
    "showInfoBox": false,
  },
  "onAddPoint": [Function],
  "onDrag": [Function],
  "onPointDelete": [Function],
  "onToggleInfoBox": [Function],
  "polylineOptions": Object {
    "strokeColor": "#4bdded",
    "strokeOpacity": 0.85,
    "strokeWeight": 4,
  },
  "render": [Function],
  "renderMarkers": [Function],
  "renderPolyline": [Function],
  "swapPoints": [Function],
  "uuidV4": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
    ],
  },
}
`;

exports[`test rendering JSX.Element 1`] = `
RouteService {
  "_componentUpdate": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
    ],
  },
  "_markers": Array [
    Object {
      "draggable": true,
      "id": "1",
      "name": "Test Marker 1",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
    Object {
      "draggable": true,
      "id": "2",
      "name": "Test Marker 2",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
    Object {
      "draggable": true,
      "id": "3",
      "name": "Test Marker 3",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": false,
    },
  ],
  "defaultPoint": Object {
    "draggable": true,
    "id": "",
    "name": "",
    "onDrag": [Function],
    "position": Object {
      "lat": 0,
      "lng": 0,
    },
    "showInfoBox": false,
  },
  "onAddPoint": [Function],
  "onDrag": [Function],
  "onPointDelete": [Function],
  "onToggleInfoBox": [Function],
  "polylineOptions": Object {
    "strokeColor": "#4bdded",
    "strokeOpacity": 0.85,
    "strokeWeight": 4,
  },
  "render": [Function],
  "renderMarkers": [Function],
  "renderPolyline": [Function],
  "swapPoints": [Function],
  "uuidV4": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
    ],
  },
}
`;

exports[`toggle markers babble 1`] = `
RouteService {
  "_componentUpdate": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
    ],
  },
  "_markers": Array [
    Object {
      "draggable": true,
      "id": "1",
      "name": "Test Marker 1",
      "onDrag": [Function],
      "position": Object {
        "lat": 0,
        "lng": 0,
      },
      "showInfoBox": true,
    },
  ],
  "defaultPoint": Object {
    "draggable": true,
    "id": "",
    "name": "",
    "onDrag": [Function],
    "position": Object {
      "lat": 0,
      "lng": 0,
    },
    "showInfoBox": false,
  },
  "onAddPoint": [Function],
  "onDrag": [Function],
  "onPointDelete": [Function],
  "onToggleInfoBox": [Function],
  "polylineOptions": Object {
    "strokeColor": "#4bdded",
    "strokeOpacity": 0.85,
    "strokeWeight": 4,
  },
  "render": [Function],
  "renderMarkers": [Function],
  "renderPolyline": [Function],
  "swapPoints": [Function],
  "uuidV4": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
      Array [],
    ],
  },
}
`;
